source/guides/deprecations/1.11/instance-initializers.md

```diff
@@ -0,0 +1,118 @@
+important for people to get access to that container to make objects
+available to the app at runtime. To do so, we introduced the initializer
+API.
+
+For example, Ember Data uses this API to ensure that all of your
+controllers and routes have access to the store object.
+
+As we have been working on the FastBoot effort to allow Ember apps to
+run on the server, we realized that the assumption that there would only
+be one instance of the app running at once no longer held.
+
+In particular, we realized that "app boot" was actually two distinct
+phases:
+
+1. Loading all of your application code and registering objects with the
+   application (such as the store in the Ember Data example)
+2. Instantiating objects when the app starts running
+
+When the application boots up for the first time, we need to run Step 1
+above, but it would be wasteful to run it for every request going to
+your FastBoot server.
+
+We also realized that a huge number of real-world initializers were
+simply registering classes and injections, while a smaller fraction of
+initializers are doing work that truly needs to be done per run.
+
+By separating out these two kinds of work, we could improve the
+performance of cases where a single application was used multiple times,
+each with a fresh set of state.
+
+This applies to FastBoot, of course, but is also applies to integration
+tests. Today, integration tests spend time executing code in
+initializers that don't change between test runs, simply because of the
+possibility that the initializer may also set up instances.
+
+Thankfully, the `initializer` infrastructure already clearly marks the
+boot-up phase of your application (both in your own code and in
+add-ons), so we only need to make a small tweak to `initializer`s for
+FastBoot to work.
+
+This change makes the difference between initializers that set up code
+and initializers that set up instances explicit.
+
+The old `App.initializer` API now receives a `registry`, which can be
+freely used to set up code using the `register` API and the various
+`inject` APIs.
+
+A new `App.instanceInitializer` API receives an `ApplicationInstance`
+object. When migrating code, the most important property on the
+`ApplicationInstance` is `container`, which you can use to get new
+instances and set state on them.
+
+Migrating code to the new structure is fairly mechanical. Change code
+that looks like:
+
+```js
+App.initializer({
+  name: "clock",
+
+  initialize: function(container) {
+    container.register("clock:main", Clock);
+    var clock = container.lookup("clock:main");
+    clock.setStartTime(Date.now());
+  }
+});
+```
+
+To:
+
+```js
+App.initializer({
+  name: "clock",
+
+  initialize: function(registry) {
+    registry.register("clock:main", Clock);
+  }
+});
+
+App.instanceInitializer({
+  name: "clock",
+
+  initialize: function(instance) {
+    var clock = instance.container.lookup("clock:main");
```

Comment:
> What is the footgun you are concerned about?

> A brief search of GitHub (https://github.com/search?utf8=%E2%9C%93&q=initializer+container.lookup&type=Code&ref=searchresults) shows many,

that search is extremely scary, for example each occurrence of `lookup('store:main')` (their are many) in that search fails to specific its initializer ordering dependency as after ember-data, which means it is entirely luck, that their application boots. Any change to the initializer dep graph can result in a topology change that may result in a broken app.

So, this means eager instantiating during initializers destabilizes application boot.

An example: 2 independent initializers that both appear reasonable, do not work together.

*Please note, the most problematic cases are when these initializers exist in add-ons, but the issue still remains without add-ons.*

```js
// initializer1.js
container.inject('router:main', 'km', 'service:kiss-metrics')
container.lookup('router:main').on('didtransition', function() { this.km.track(..) });

// initializer2.js
container.inject('router:main', 'google', 'service:google-analtyics')
container.lookup('router:main').on('didtransition', function() { this.google.track(..) });
```

depending on the initialization order, either google analytics or kiss metrics will be broken.

proposed solution:

```js
// initializer1.js
container.inject('router:main', 'km', 'service:kiss-metrics')
container.onLookup('router:main', function(router) {
  router.on('didTransition', function() { this.km.track(..) })
});

// initializer2.js
container.inject('router:main', 'google', 'service:google-analtyics')
container.onLookup('router:main', function(router) {
  router.on('didTransition', function() { this.google.track(..) })
});```